File: src/validate.ts
import { readFile } from 'node:fs/promises';
import { join } from 'node:path';
import { Ajv, type ValidateFunction } from 'ajv';
import addMetaSchema2020Import from 'ajv/dist/refs/json-schema-2020-12/index.js';
import addFormatsPlugin, { type FormatsPluginOptions } from 'ajv-formats';
import { createRequire } from 'node:module';
import type { NormalizedDataBundle } from './types/index.js';

const require = createRequire(import.meta.url);
const ajv = new Ajv({ allErrors: true, strict: false });
if (typeof addMetaSchema2020Import === 'function') {
  (addMetaSchema2020Import as unknown as (this: Ajv, $data?: boolean) => Ajv).call(ajv);
} else {
  const metaSchemaFn = (addMetaSchema2020Import as {
    default?: (this: Ajv, $data?: boolean) => Ajv;
  }).default;
  if (typeof metaSchemaFn === 'function') {
    metaSchemaFn.call(ajv);
  }
}
const metaSchema202012 = require('ajv/dist/refs/json-schema-2020-12/schema.json');
if (!ajv.getSchema('https://json-schema.org/draft/2020-12/schema')) {
  ajv.addMetaSchema(metaSchema202012);
}
const addFormats = addFormatsPlugin as unknown as (
  ajv: Ajv,
  options?: FormatsPluginOptions
) => Ajv;
addFormats(ajv);

const validatorCache = new Map<string, ValidateFunction>();

async function loadValidator (schemaPath: string): Promise<ValidateFunction> {
  if (validatorCache.has(schemaPath)) {
    return validatorCache.get(schemaPath)!;
  }
  const schema = JSON.parse(await readFile(schemaPath, 'utf8'));
  const validator = ajv.compile(schema);
  validatorCache.set(schemaPath, validator);
  return validator;
}

function ensureArray (data: unknown, label: string): asserts data is unknown[] {
  if (!Array.isArray(data)) {
    throw new Error(`Expected ${label} to be an array.`);
  }
}

export async function validateArray (schemaPath: string, data: unknown, label: string) {
  ensureArray(data, label);
  const validator = await loadValidator(schemaPath);
  for (const [index, entry] of data.entries()) {
    if (!validator(entry)) {
      const message = ajv.errorsText(validator.errors, { dataVar: `${label}[${index}]` });
      throw new Error(message);
    }
  }
}

const defaultSchemaDir = join(process.cwd(), 'schemas');

export async function validateNormalizedBundle (
  bundle: NormalizedDataBundle,
  schemaDir: string = defaultSchemaDir
) {
  await Promise.all([
    validateArray(join(schemaDir, 'manufacturer.json'), bundle.manufacturers, 'manufacturers'),
    validateArray(join(schemaDir, 'ship.json'), bundle.ships, 'ships'),
    validateArray(join(schemaDir, 'ship_variant.json'), bundle.ship_variants, 'ship_variants'),
    validateArray(join(schemaDir, 'item.json'), bundle.items, 'items'),
    validateArray(join(schemaDir, 'hardpoint.json'), bundle.hardpoints, 'hardpoints'),
    validateArray(join(schemaDir, 'item_stats.json'), bundle.item_stats, 'item_stats'),
    validateArray(join(schemaDir, 'ship_stats.json'), bundle.ship_stats, 'ship_stats'),
    validateArray(join(schemaDir, 'installed_item.json'), bundle.installed_items, 'installed_items'),
    validateArray(join(schemaDir, 'locale.json'), bundle.locales, 'locales')
  ]);
}

File: internal/validate/validate.go
package validate

import (
	"context"
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"sync"

	jsonschema "github.com/santhosh-tekuri/jsonschema/v6"

	"github.com/ArisCorporation/sc-goetl/internal/model"
)

var (
	schemaCache sync.Map
)

// Bundle validates the normalized structures against JSON Schemas.
func Bundle(ctx context.Context, bundle model.NormalizedDataBundle, schemaDir string) error {
	tasks := []struct {
		File string
		Data any
	}{
		{File: "manufacturer.json", Data: bundle.Manufacturers},
		{File: "ship.json", Data: bundle.Ships},
		{File: "ship_variant.json", Data: bundle.ShipVariants},
		{File: "item.json", Data: bundle.Items},
		{File: "hardpoint.json", Data: bundle.Hardpoints},
		{File: "item_stats.json", Data: bundle.ItemStats},
		{File: "ship_stats.json", Data: bundle.ShipStats},
		{File: "installed_item.json", Data: bundle.InstalledItems},
		{File: "locale.json", Data: bundle.Locales},
	}

	for _, task := range tasks {
		path := filepath.Join(schemaDir, task.File)
		if err := validateArray(ctx, path, task.Data); err != nil {
			return fmt.Errorf("validate %s: %w", task.File, err)
		}
	}
	return nil
}

func validateArray(ctx context.Context, schemaPath string, data any) error {
	if _, err := os.Stat(schemaPath); err != nil {
		return err
	}
	schema, err := loadSchema(schemaPath)
	if err != nil {
		return err
	}
	generic, err := toInterface(data)
	if err != nil {
		return err
	}
	arr, ok := generic.([]any)
	if !ok {
		return fmt.Errorf("expected array input for %s", schemaPath)
	}
	for idx, item := range arr {
		if err := schema.Validate(item); err != nil {
			return fmt.Errorf("%s[%d]: %w", schemaPath, idx, err)
		}
	}
	return nil
}

func loadSchema(path string) (*jsonschema.Schema, error) {
	if value, ok := schemaCache.Load(path); ok {
		return value.(*jsonschema.Schema), nil
	}
	compiler := jsonschema.NewCompiler()
	compiler.DefaultDraft(jsonschema.Draft2020)

	bytes, err := os.ReadFile(path)
	if err != nil {
		return nil, err
	}
	var doc any
	if err := json.Unmarshal(bytes, &doc); err != nil {
		return nil, err
	}
	if err := compiler.AddResource(path, doc); err != nil {
		return nil, err
	}
	schema, err := compiler.Compile(path)
	if err != nil {
		return nil, err
	}
	schemaCache.Store(path, schema)
	return schema, nil
}

func toInterface(data any) (any, error) {
	bytes, err := json.Marshal(data)
	if err != nil {
		return nil, err
	}
	var v any
	if err := json.Unmarshal(bytes, &v); err != nil {
		return nil, err
	}
	return v, nil
}
